mm2src/coins/nft/nft_structs.rs

```diff
@@ -129,13 +133,55 @@ impl fmt::Display for ContractType {
 }
 
 #[derive(Clone, Debug, Default, Deserialize, Serialize)]
-pub(crate) struct UriMeta {
+pub(crate) struct UriMetaFromStr {
     pub(crate) image: Option<String>,
+    /// the same as image, some tokens use **image_url** name for **image** field
+    /// but sometimes token_uri contains both`image_url` and `image`
+    pub(crate) image_url: Option<String>,
     #[serde(rename(deserialize = "name"))]
     pub(crate) token_name: Option<String>,
-    description: Option<String>,
-    attributes: Option<Json>,
-    animation_url: Option<String>,
+    pub(crate) description: Option<String>,
+    pub(crate) attributes: Option<Json>,
+    pub(crate) animation_url: Option<String>,
+    pub(crate) external_url: Option<String>,
+    pub(crate) image_details: Option<Json>,
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+pub(crate) struct UriMeta {
+    pub(crate) image: Option<String>,
+    pub(crate) token_name: Option<String>,
+    pub(crate) description: Option<String>,
+    pub(crate) attributes: Option<Json>,
+    pub(crate) animation_url: Option<String>,
+    pub(crate) external_url: Option<String>,
+    pub(crate) image_details: Option<Json>,
+}
```

Comment:
I might be missing something, but can't these 2 structs be only one?

Reply:
I use `UriMeta` as final structure for `uri_meta` field in `Nft`. 
As I noted on doc comments `image` is the same as `image_url`, just sometimes json from `token_uri` or string from `metadata` field contain `image`, sometimes they(nft creators) call it as `image_url`. Sometimes there are both image and image_url fileds. I created `UriMetaFromStr` to be able to collect both possible variants and not to have extra `image` filed later in `UriMeta`. 

After merging `UriMeta` will have just `image`, so GUI or cli users wont need to check `image` and then `image_url`, they will have the final `image`, bcz we already checked possible nulls

Reply:
> After merging UriMeta will have just image, so GUI or cli users wont need to check image and then image_url, they will have the final image, bcz we already checked possible nulls

This can be handled before returning the value to GUI clients I think. Like you can add field `raw_image_url` which contains the original value, and `image_url` for the computed/updated/proper value.

Reply:
> This can be handled before returning the value to GUI clients I think. Like you can add field raw_image_url which contains the original value, and image_url for the computed/updated/proper value.

I need to build the structure from `token_uri` and `from metadata`, so I need to keep `image` and `image_url` names somewhere.
If you want to keep one structure, then lets have three optional params: image, image_url, merged_image

Reply:
You can use `rename` attribute for this.

https://serde.rs/field-attrs.html#rename

Reply:
The whole idea was not to keep two image fields, if you want to keep the two I will change it.

Reply:
Having 2 image fields is much better than duplicating whole type. You can take the raw data in `raw_image_url`, do the necessary calculations and store the proper value in `image_url`, then skip returning `raw_image_url` one to the gui clients.

Reply:
> then skip returning raw_image_url one to the gui clients.

Im afraid that `skip_serializing` will lead to problems, bcz serialize nft into detail_json and from json into nft in database implementations. Once we deserialize structure back from database, the field will be none, then why we keep this field.

Ok not so problems may be, just `raw_image` will always be None.

Reply:
If it makes too much confusion, maybe return the value to the GUI clients too. Maybe it could be even better to give them the original value too?

Reply:
> If it makes too much confusion, maybe it could be even better to give them the original value too?

Yeah, lets do this, it is better.

Reply:
Done. Thanks for this long conversation @shamardy @ozkanonur , it was important.

I left only `UriMeta` structure and I build it using `get_uri_meta` function.

I added kinda long doc comments for `struct UriMeta` and `merge_from` function, but its worth adding to make it clearer what is going on.

Also previously I was using `#[serde(rename(deserialize = "name"))]` annotation for `pub(crate) token_name: Option<String>,` it was wrong, as I constantly need to serialize and de serialize `UriMeta` structure: 
deserialize when we create it from token_uri response or from `metadata` field,
serialize it when we save it as a part of `detail_json` in database, 
deserialize it again to get the object from database.

So, I put `#[serde(rename = "image")]` for `raw_image_url` and `#[serde(rename = "name")]` for `token_name`. Without this I can have `null` in responses, when we get the data from DB.

One more note, `merge_from` function doesnt change `raw_image_url` which is originally "image" in `token_uri` and in`metadata`. Despite the fact that often nft creators use only `image` name, lets try to update `image_url` field. 
So probably we will have two Some in the end, `raw_image_url` and `image_url` (and I will need to provide a note in smk docs that `image` in responses is kinda `raw_image_url` and its better to look at `image_url`), but yeah its better to keep untouched version and updated version